### PR TITLE
asserts: start supporting optional primary keys in fs backend, assemble and signing

### PR DESCRIPTION
**Pull request by pedronis**
We have come to the conclusion that we might need to support adding optional primary key headers to preexisting assertion types, to allow discriminating and avoiding clashing when for example we allow signing delegation for the type.

This must be done in such a way that assertions using just the default values are still accessible if there's a snapd revert. In particular a system not using non-default values should be able to revert without issues.

This starts adding support for optional primary key headers concentrating on the basics of the assemble functions and signing and support in the filesystem backend, the strategy there is to make it so that assertions using the default values are stored at the same paths as before. While assertions using non-default values are stored at non clashing paths.

### CODE REVIEW

File: asserts/asserts.go [hunk 1]
```diff
@@ -1066,10 +1103,21 @@ func assembleAndSign(assertType *AssertionType, headers map[string]interface{},
...
 	}
 	for _, primKey := range assertType.PrimaryKey {
-		if _, err := checkPrimaryKey(finalHeaders, primKey); err != nil {
+		defl := assertType.OptionalPrimaryKeyDefaults[primKey]
+		_, ok := finalHeaders[primKey]
+		if !ok && defl != "" {
+			// optional but expected to be set in headers
+			// in the result assertion
+			finalHeaders[primKey] = defl
+			continue
```
Comment (mardy) [inline asserts/asserts.go:1112]: It's not clear to me why this `continue` is here, but on the other hand I don't understand what the rest of code code does either, so I'm not sure you can explain this in a short comment. But if it's feasible, please do :-)

**Review by mardy — APPROVED** [review]
LGTM, one question inline.

File: asserts/asserts.go [hunk 2]
```diff
@@ -57,11 +57,39 @@ type AssertionType struct {
 	// PrimaryKey holds the names of the headers that constitute the
 	// unique primary key for this assertion type.
 	PrimaryKey []string
+	// OptionalPrimaryKeyDefaults holds the default values for
+	// optional primary key headers.
+	// Optional primary key headers can be added to types defined
+	// in previous versions of snapd, as long as they are added at
+	// the end of the old primary key together with a default value set in
+	// this map.
```
Comment (bboozzoo) [inline asserts/asserts.go:65]: Perhaps it's implicit, but maybe it should mention that there needs to be a default value for each primary key that follows the first key with a default value?

**Review by bboozzoo — APPROVED** [review]
LGTM, do we need this for memory backstore too?

**Comment by bboozzoo** [comment]
This branch needs a merge from master, there are conflicts now.

**Pull request merged**